internal/tools/restart.go

```diff
@@ -0,0 +1,129 @@
+package tools
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/mark3labs/mcp-go/mcp"
+	"github.com/mark3labs/mcp-go/server"
+	v1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+	pipelineclient "github.com/tektoncd/pipeline/pkg/client/injection/client"
+	pipelineruninformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1/pipelinerun"
+	taskruninformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1/taskrun"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func restartPipelineRun() server.ServerTool {
+	return server.ServerTool{
+		Tool: mcp.NewTool("restart_pipelinerun",
+			mcp.WithDescription("Restart a PipelineRun"),
+			mcp.WithString("name", mcp.Required(),
+				mcp.Description("Name or Reference of the PipelineRun to restart"),
+			),
+			mcp.WithString("namespace",
+				mcp.Description("Namespace where the PipelineRun is located"),
+				mcp.DefaultString("default"),
+			),
+			// TODO add "parameters" objects
+		),
+		Handler: handlerRestartPipelineRun,
+	}
+}
+
+func handlerRestartPipelineRun(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+	name, ok := request.Params.Arguments["name"].(string)
+	if !ok {
+		return mcp.NewToolResultError("namespace must be a string"), nil
+	}
+	namespace, ok := request.Params.Arguments["namespace"].(string)
+	if !ok {
+		return mcp.NewToolResultError("namespace must be a string"), nil
+	}
+
+	pipelinerunInformer := pipelineruninformer.Get(ctx)
+	pipelineclientset := pipelineclient.Get(ctx)
+
+	usepr, err := pipelinerunInformer.Lister().PipelineRuns(namespace).Get(name)
+	if err != nil {
+		return mcp.NewToolResultErrorFromErr(fmt.Sprintf("Failed to get PipelineRun %s/%s", namespace, name), err), nil
+	}
+	pr := &v1.PipelineRun{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "tekton.dev/v1",
+			Kind:       "PipelineRun",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace,
+			GenerateName: name + "-",
+		},
+		Spec: usepr.Spec,
+	}
+	pr.Spec.Status = ""
+	if len(usepr.ObjectMeta.GenerateName) > 0 {
+		pr.ObjectMeta.GenerateName = usepr.ObjectMeta.GenerateName
+	}
+
+	if _, err := pipelineclientset.TektonV1().PipelineRuns(namespace).Create(ctx, pr, metav1.CreateOptions{}); err != nil {
+		return mcp.NewToolResultErrorFromErr(fmt.Sprintf("Failed to create PipelineRun %s/%s", namespace, name), err), nil
+	}
+
+	return result(fmt.Sprintf("Restarting pipelinerun %s in namespace %s", name, namespace)), nil
```

Comment:
Should the `name` here be the name of the pipelinerun started, or of the one we are restarting :thinking: or perhaps both

Reply:
@jbpratt most likely both. "Restarting pipelinerun {oldname} as {nemname} in namespace {namespace}".

Reply:
Updated!

Reply:
Okay it is fixed now, actually this time

```
Restarting taskrun hello-world-run-run-hello as hello-world-run-run-hello-md5lf in namespace default
```